crates/hir-ty/src/infer.rs

```diff
@@ -649,83 +653,138 @@ impl InferenceResult {
-    pub fn pattern_types(&self) -> impl Iterator<Item = (PatId, &Ty)> {
-        self.type_of_pat.iter()
+    pub fn pattern_types(&self) -> impl Iterator<Item = (PatId, Ty<'db>)> {
+        self.type_of_pat.iter().map(|(k, v)| (k, *v))
     }
 
     // This method is consumed by external tools to run rust-analyzer as a library. Don't remove, please.
-    pub fn binding_types(&self) -> impl Iterator<Item = (BindingId, &Ty)> {
-        self.type_of_binding.iter()
+    pub fn binding_types(&self) -> impl Iterator<Item = (BindingId, Ty<'db>)> {
+        self.type_of_binding.iter().map(|(k, v)| (k, *v))
     }
 
     // This method is consumed by external tools to run rust-analyzer as a library. Don't remove, please.
-    pub fn return_position_impl_trait_types(&self) -> impl Iterator<Item = (ImplTraitIdx, &Ty)> {
-        self.type_of_rpit.iter()
+    pub fn return_position_impl_trait_types(
+        &self,
+    ) -> impl Iterator<Item = (ImplTraitIdx<'db>, Ty<'db>)> {
+        self.type_of_rpit.iter().map(|(k, v)| (k, *v))
     }
 }
 
-impl Index<ExprId> for InferenceResult {
-    type Output = Ty;
+impl<'db> Index<ExprId> for InferenceResult<'db> {
+    type Output = Ty<'db>;
 
-    fn index(&self, expr: ExprId) -> &Ty {
-        self.type_of_expr.get(expr).unwrap_or(&self.standard_types.unknown)
+    fn index(&self, expr: ExprId) -> &Ty<'db> {
+        self.type_of_expr.get(expr).unwrap_or(&self.error_ty)
     }
 }
 
-impl Index<PatId> for InferenceResult {
-    type Output = Ty;
+impl<'db> Index<PatId> for InferenceResult<'db> {
+    type Output = Ty<'db>;
 
-    fn index(&self, pat: PatId) -> &Ty {
-        self.type_of_pat.get(pat).unwrap_or(&self.standard_types.unknown)
+    fn index(&self, pat: PatId) -> &Ty<'db> {
+        self.type_of_pat.get(pat).unwrap_or(&self.error_ty)
     }
 }
 
-impl Index<ExprOrPatId> for InferenceResult {
-    type Output = Ty;
+impl<'db> Index<ExprOrPatId> for InferenceResult<'db> {
+    type Output = Ty<'db>;
 
-    fn index(&self, id: ExprOrPatId) -> &Ty {
-        self.type_of_expr_or_pat(id).unwrap_or(&self.standard_types.unknown)
+    fn index(&self, id: ExprOrPatId) -> &Ty<'db> {
+        match id {
+            ExprOrPatId::ExprId(id) => &self[id],
+            ExprOrPatId::PatId(id) => &self[id],
+        }
     }
 }
 
-impl Index<BindingId> for InferenceResult {
-    type Output = Ty;
+impl<'db> Index<BindingId> for InferenceResult<'db> {
+    type Output = Ty<'db>;
 
-    fn index(&self, b: BindingId) -> &Ty {
-        self.type_of_binding.get(b).unwrap_or(&self.standard_types.unknown)
+    fn index(&self, b: BindingId) -> &Ty<'db> {
+        self.type_of_binding.get(b).unwrap_or(&self.error_ty)
     }
 }
 
 #[derive(Debug, Clone)]
-struct InternedStandardTypesNextSolver<'db> {
-    unit: crate::next_solver::Ty<'db>,
-    never: crate::next_solver::Ty<'db>,
-    i32: crate::next_solver::Ty<'db>,
-    f64: crate::next_solver::Ty<'db>,
+struct InternedStandardTypes<'db> {
```

Comment:
I used these in more places, to avoid re-interning of common types, and removed it from `InferenceResult` (moved to `InferenceContext`), except for the error type which can't be moved easily.